### ISSUE TITLE
showteam page doesn't work with group leaders

### ISSUE THREAD

**Issue by yuliu** [mybb/mybb#3785]
This is a v1.9 problem.

The `showteam` page outputs an `Uncaught TypeError: Argument 1 passed to MyBB\Twig\Extensions\UrlExtension::getProfileLink() must be of the type integer, null given, called in...` error when the board has group leaders to show.

Probably caused by neglect during theme migration to Twig. I think I can fix it.

BTW, for the 1.9 development branch, should I wait for the code being rebased before I can pull a request? New to git rebase and current MyBB 1.9 code rebase.

**Comment by euantorano** [comment]
Feel free to open PRs - the rebase is a long and arduous task that I’m working on as and when possible. I prioritise PRs otherwise they end up conflicted.

**Comment by yuliu** [comment]
I'm not good at CSS scripting, and thus I really can't fix most of the style issues posted here. Better mention, maybe justinsoltesz and effone and others, to check on those problems.

**Comment by yuliu** [comment]
Looks like the doubled `&&amp;` in some URLs are caused by double escaping, one by MyBB's function `htmlspecialchars_uni()`, and the other by Twig'a auto-escaping that defaults to `html`.

I didn't find any MyBB's environment settings for Twig for now, so variables fetched in `{{ }}` will be escaped by Twig by default. Or we should remove MyBB's `htmlspecialchars_uni()` on those `get_*_link()`.

See https://twig.symfony.com/doc/2.x/templates.html#html-escaping and https://twig.symfony.com/doc/2.x/api.html#environment-options .

**Comment by euantorano** [comment]
We configure the environment here: https://github.com/mybb/mybb/blob/develop/1.9/inc/src/Twig/ServiceProvider.php#L67

The `get_*_link()` functions should be changed IMO though - one of the biggest advantages of the move to Twig is the auto escaping.

**Comment by yuliu** [comment]
Looks there are a lot of code to be examined for `get_*_link()` functions. Besides this. I have found other possible/questionable issues on (have been listed in the main issue):

- Avatar for guest/non-user should be outside a `<a href>` link.

- Some controls that are originally permissions are calculated in the template. Shouldn't we have them in PHP code? Eg, the `newthread` button in `forumdisplay`. There are a few questionable controls like this.

**Comment by euantorano** [comment]
It might be nice to have the `get_*_link()` functions available within Twig, but I'm not sure.

- Agreed. Not sure what would be best for that, perhaps another argument to the function like `bool $renderLink`?
- In some ways it's nice to have permission checks inside templates, but I don't mind too much either way.

**Comment by yuliu** [comment]
`get_*_link()` might be used from outside the template, not sure.

Sorry, seem I didn't make my words clear with the other two issues.

- We'd better just use `<a href=profile_link><user_avatar></a>` for existing users, and  `<a href=profile_link><default_user_avatar></a>`. The `default_user_avatar` will be shown with empty parameter on calling `renderAvatar`.

- For permissions that are defined in the forum settings and user groups, better leave the job with PHP script, but not the template. Twig will mostly do the template job rather than the PHP script' job. Otherwise, the template will be incompatible from version to version.

**Comment by euantorano** [comment]
> get_*_link() might be used from outside the template, not sure.

Yes, it will be (such as for RSS). It would be nice to have a class or something that can generate URLs for all sorts of pages (eg: `calendar`, `forum`, `thread`, `profile`). At the moment certain links (eg: `calendar` => `{$mybb->settings['bburl']}/calendar.php`) are just hardcoded, whereas it would be nice to do `url('calendar')` so that if/when we move to a route based system (where `showthread.php?tid=X` may be `thread/SLUG`) we can easily move across by simply changing the class that generates the URLs.

Said class would then have a Twig extension that can be used in templates, and a helper function to be used in PHP.

> We'd better just use <a href=profile_link><user_avatar></a> for existing users, and <a href=profile_link><default_user_avatar></a>. The default_user_avatar will be shown with empty parameter on calling renderAvatar.

Apologies, it's been a while since I wrote the code/template for the avatar stuff and I couldn't remember how it worked. I think I understand what you're meaning now.

> For permissions that are defined in the forum settings and user groups, better leave the job with PHP script, but not the template. Twig will mostly do the template job rather than the PHP script' job. Otherwise, the template will be incompatible from version to version.

Ok, so instead of doing something like:

```twig
{% if thread.canfullreply %}
                        <a href="newreply.php?tid={{ thread.tid }}" class="button button--medium button--full-width">
                            {{ include('partials/icon.twig', {icon: 'comment', class: 'button__icon'}, with_context = false) }}
                            <span class="button__text">{{ lang.new_reply }}</span>
                        </a>
                    {% elseif thread.canmodreply %}
                        <a href="newreply.php?tid={{ thread.tid }}" class="button button--medium button--danger">
                            {{ include('partials/icon.twig', {icon: 'lock', class: 'button__icon'}, with_context = false) }}
                            <span class="button__text">{{ lang.thread_closed }}</span>
                        </a>
                    {% else %}
                        <a href="showthread.php?tid={{ thread.tid }}" class="button button--medium button--danger">
                            {{ include('partials/icon.twig', {icon: 'lock', class: 'button__icon'}, with_context = false) }}
                            <span class="button__text">{{ lang.thread_closed }}</span>
                        </a>
                    {% endif %}
```

You'd instead pass a bunch of variables like `can_reply`, `reply_text`, similar to how we do it in 1.8?

**Comment by yuliu** [comment]
Get it. The URL thing seems requiring a big job.

Your point makes sense, too. The template will become bloated. My original method for displaying avatar for existing users and guest/deleted is just using more `if` controls. Look like we should use a proxy function or something similar, as we've discussed on Discord.

**Comment by yuliu** [comment]
Two private message menu items are added back, `empty folder` and `export messages`. Icons might need review.

I'd like to ping @justinsoltesz here, again. Could you please have a check on the minor problems listed under the **Style** section? I've looked at them but have no idea how to fix them. Besides the above mentioned icon thing, is it possible and suitable to have some `sub-menu-item` classes with indent for listing PM folders under an item called `Folders`?
